cpp/oneapi/dal/table/backend/homogen_table_impl.hpp

```diff
@@ -20,13 +20,84 @@
 
 namespace oneapi::dal::backend {
 
+enum class alloc_kind {
+    host, /// Non-USM pointer allocated on host
+    usm_host, /// USM pointer allocated by sycl::alloc_host
+    usm_device, /// USM pointer allocated by sycl::alloc_device
+    usm_shared /// USM pointer allocated by sycl::alloc_shared
+};
+
+#ifdef ONEDAL_DATA_PARALLEL
+inline alloc_kind alloc_kind_from_sycl(sycl::usm::alloc alloc) {
+    using error_msg = dal::detail::error_messages;
+    switch (alloc) {
+        case sycl::usm::alloc::host: return alloc_kind::usm_host;
+        case sycl::usm::alloc::device: return alloc_kind::usm_device;
+        case sycl::usm::alloc::shared: return alloc_kind::usm_shared;
+        default: throw invalid_argument{ error_msg::unsupported_usm_alloc() };
+    }
+}
+#endif
+
+#ifdef ONEDAL_DATA_PARALLEL
+inline sycl::usm::alloc alloc_kind_to_sycl(alloc_kind kind) {
+    switch (kind) {
+        case alloc_kind::usm_host: return sycl::usm::alloc::host;
+        case alloc_kind::usm_device: return sycl::usm::alloc::device;
+        case alloc_kind::usm_shared: return sycl::usm::alloc::shared;
+        case alloc_kind::host: return sycl::usm::alloc::unknown;
+        default: //
+            ONEDAL_ASSERT("Unsupported alloc_kind");
+    }
+}
+#endif
+
+inline bool alloc_kind_requires_copy(alloc_kind src_alloc_kind, alloc_kind dst_alloc_kind) {
+#ifdef ONEDAL_DATA_PARALLEL
+    switch (dst_alloc_kind) {
+        case alloc_kind::host: //
+            return (src_alloc_kind == alloc_kind::usm_device);
+        case alloc_kind::usm_host: //
+            return (src_alloc_kind == alloc_kind::host) || //
+                   (src_alloc_kind == alloc_kind::usm_device);
+        case alloc_kind::usm_device: //
+            return (src_alloc_kind == alloc_kind::host) || //
+                   (src_alloc_kind == alloc_kind::usm_host);
+        case alloc_kind::usm_shared: //
+            return (src_alloc_kind != alloc_kind::usm_shared);
```

Comment:
Why it is required to copy from `usm_host` in this case?

Reply:
The same considerations mentioned [in the previous comment](https://github.com/oneapi-src/oneDAL/pull/1488#discussion_r595822344).